blogs/posts/2024-02-11_sankey_plot.qmd

```diff
@@ -0,0 +1,565 @@
+    summarise(
+      flow = n_distinct(participant_id, na.rm = T),
+      .by = c(from, to)
+    )
+)
+
+# view our data
+df_flows |> 
+  reactable(defaultPageSize = 5)
+```
+
+# Sankey plot
+
+We now have a neat little summary of movements of participants between the milestones in our recruitment pathway. However, this 'tidy' data isn't the format required by [plotly](https://plotly.com/r/sankey-diagram/), so the next steps are to prepare it ready for plotting.
+
+## Preparing for plotly
+
+Plotly expects to be fed two sets of data, nodes and edges:
+
+1.  Nodes - these are the milestones we have in our `from` and `to` fields,
+
+2.  Edges - these are the flows that occur between nodes, the `flow` in our table.
+
+It is possible to extract this data by hand, however, I found using the [tidygraph](https://tidygraph.data-imaginist.com) package much easier and more convenient.
+
+```{r}
+df_sankey <- df_flows |> 
+  # convert our flows data to a tidy graph object
+  as_tbl_graph()
+
+```
+
+The tidygraph package splits our data into nodes and edges. We can selectively work on each by 'activating' them - here is the nodes list:
+
+```{r}
+df_sankey |> 
+  activate(what = 'nodes') |> 
+  as_tibble() |> 
+  reactable(defaultPageSize = 5)
+```
+
+You can see each unique node name listed. The row numbers for these nodes are used as reference ids in the edges object:
+
+```{r}
+df_sankey |> 
+  activate(what = 'edges') |> 
+  as_tibble() |> 
+  reactable(defaultPageSize = 5)
+```
+
+We now have enough information to generate our Sankey.
+
+First we extract our nodes and edges to separate data frames then convert the id values to be zero-based (starts at 0) as this is what plotly is expecting. To do this is as simple as subtracting 1 from the value of the ids.
+
+Finally we pass these two dataframes to plotly's `node` and `link` function inputs to generate the plot.
+
+```{r}
+#| code-fold: true
+#| fig-cap: Our first sankey
+#| fig-cap-location: top
+
+# extract the nodes to a dataframe
+nodes <- df_sankey |> 
+  activate(nodes) |> 
+  data.frame() |> 
+  mutate(
+    id = row_number() -1
+  )
+
+# extract the edges to a dataframe
+edges <- df_sankey |> 
+  activate(edges) |> 
+  data.frame() |> 
+  mutate(
+    from = from - 1,
+    to = to - 1
+  )
+
+# plot our sankey
+plot_ly(
+  # setup
+  type = 'sankey',
+  orientation = 'h',
+  arrangement = 'snap',
+  
+  # use our node data
+  node = list(
+    label = nodes$name
+  ),
+  
+  # use our link data
+  link = list(
+    source = edges$from,
+    target = edges$to,
+    value = edges$flow
+  )
+)
+```
+
+Not bad!
+
+We can see the structure of our Sankey now. Can you see the relative proportions of participants who did or didn't respond to our first invite? Marvel at how those who responded to the first invite flow into our final outcome. How about those who didn't respond to the first invitation go on to receive a second invite? This looks functional.
```

Comment:
I wonder if you should specifically invite people to hover their cursor over the nodes and edges, just in case they haven't clocked that it's interactive!